### DISCUSSION TITLE
How can i have Client_id being returned from OAuth2PasswordRequestForm

### DISCUSSION THREAD

**Discussion by mittal-umang** [fastapi/fastapi#9029] [Questions] [answered]
I'm currently trying to implement OAuth2 for a project and the OAuth2 provider being used requires the client_id while introspecting the token in headers. but i end up losing it after the `OAuth2PasswordRequestForm` i want to be able to return client_id from the function that depends on `OAuth2PasswordRequestForm` and eventually utilize client_id to make a request to the OAuth2 provider as a header. 

Any help will be appreciated.

As a work around i have concatenated the client_id with the token from the provider and passed it. and created `CustomOAuth2PasswordBearer` to which splits it into three parameters and return the token and client_id.

```Python
class CustomOAuth2PasswordBearer(OAuth2):
    def __init__(
            self,
            tokenUrl: str,
            scheme_name: Optional[str] = None,
            scopes: Optional[Dict[str, str]] = None,
            auto_error: bool = True,

    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlows(password={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)

    async def __call__(self, request: Request) -> Optional[Tuple[str, str]]:
        header: str = request.headers.get("Authorization")
        if header:
            scheme, param, client_id = header.split(" ")
        else:
            scheme, param, client_id = ["", "", ""]
        if not header or str(scheme).lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param, client_id
```

**Comment by mittal-umang** [comment]
@tiangolo 
sorry to corner you like this. but do you think this is possible?

**Comment by Alfro** [comment]
Hey @mittal-umang, how are you implementing your endpoint using ``OAuth2PasswordRequestForm``?

I took a look at the class implementation and it does seem to have a client_id as an optional field:
https://github.com/tiangolo/fastapi/blob/3b2e891917dd2739659643cedb5258e65d54934d/fastapi/security/oauth2.py#L40

**Comment by YuriiMotov** [comment, marked as answer]
This will be against the standard.
But, you can subclass from existing class and re-implement the logic as you need if you want what you are doing :)
